### PR TITLE
Make shared stacks visible again

### PR DESCRIPTION
**Pull request by AlexejPenner**
## Describe changes
I synced the /stacks and /projects/<ID>/stacks endpoints to use the same logic. Also made it so that the user filter can not be used to get someone elses private stacks.

## Pre-requisites
Please ensure you have done the following:
- [ ] I have read the **CONTRIBUTING.md** document.
- [ ] If my change requires a change to docs, I have updated the documentation accordingly.
- [ ] If I have added an integration, I have updated the [integrations](https://docs.zenml.io/component-gallery/integrations) table and the [corresponding website section](https://zenml.io/integrations).
- [ ] I have added tests to cover my changes.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Other (add details above)

### CODE REVIEW

**Comment by AlexejPenner** [comment]
@fa9r I know this is yet another reason to have one endpoint instead of two for each functionality, looking forward to when we get there :)

File: src/zenml/zen_server/routers/stacks_endpoints.py [hunk 1]
```diff
@@ -64,10 +64,14 @@ def list_stacks(
     stacks: List[StackResponseModel] = []
 
     # Get private stacks unless `is_shared` is set to True
-    if is_shared is None or not is_shared:
+    if (
+        is_shared is None
+        or not is_shared
+        and user_name_or_id == auth_context.user.id
```
Comment (fa9r) [inline src/zenml/zen_server/routers/stacks_endpoints.py:70]: This won't work for `list_stacks(user_name_or_id=None)` which is the default `list_stacks()` call right?
```suggestion
        and (not user_name_or_id or user_name_or_id == auth_context.user.id)
```

File: src/zenml/zen_server/routers/stack_components_endpoints.py [hunk 2]
```diff
@@ -73,10 +73,14 @@ def list_stack_components(
     components: List[ComponentResponseModel] = []
 
     # Get private stack components unless `is_shared` is set to True
-    if is_shared is None or not is_shared:
+    if (
+        is_shared is None
+        or not is_shared
+        and user_name_or_id == auth_context.user.id
```
Comment (fa9r) [inline src/zenml/zen_server/routers/stack_components_endpoints.py:79]:
```suggestion
        and (not user_name_or_id or user_name_or_id == auth_context.user.id)
```
same as below

File: src/zenml/zen_server/routers/projects_endpoints.py [hunk 3]
```diff
@@ -329,15 +340,25 @@ def list_project_stack_components(
...
-        is_shared=False,
-    )
-    # In case the user didn't explicitly filter for is shared == False
+    components: List[ComponentResponseModel] = []
+
+    # Get private stack components unless `is_shared` is set to True
+    if (
+        is_shared is None
+        or not is_shared
+        and user_name_or_id == auth_context.user.id
```
Comment (fa9r) [inline src/zenml/zen_server/routers/projects_endpoints.py:349]: same as below

```suggestion
        and (not user_name_or_id or user_name_or_id == auth_context.user.id)
```

File: src/zenml/zen_server/routers/projects_endpoints.py [hunk 4]
```diff
@@ -231,18 +231,29 @@ def list_project_stacks(
...
-    )
-    # In case the user didn't explicitly filter for is shared == False
+    stacks: List[StackResponseModel] = []
+
+    # Get private stacks unless `is_shared` is set to True
+    # Get private stack components unless `is_shared` is set to True
+    if (
+        is_shared is None
+        or not is_shared
+        and user_name_or_id == auth_context.user.id
```
Comment (fa9r) [inline src/zenml/zen_server/routers/projects_endpoints.py:241]: same as below

```suggestion
        and (not user_name_or_id or user_name_or_id == auth_context.user.id)
```

**Review by fa9r — CHANGES_REQUESTED** [review]
Looks good apart from one thing that I marked 4 times which is that the default `list_stacks()` might not contain any private stacks anymore with the new if condition check

File: src/zenml/zen_server/routers/stacks_endpoints.py (hunk 1 above)
Comment (AlexejPenner) [reply src/zenml/zen_server/routers/stacks_endpoints.py]: good point, changed it everywhere

File: src/zenml/zen_server/routers/stack_components_endpoints.py (hunk 2 above)
Comment (AlexejPenner) [reply src/zenml/zen_server/routers/stack_components_endpoints.py]: good point, changed it everywhere

File: src/zenml/zen_server/routers/projects_endpoints.py (hunk 4 above)
Comment (AlexejPenner) [reply src/zenml/zen_server/routers/projects_endpoints.py]: good point, changed it everywhere

File: src/zenml/zen_server/routers/projects_endpoints.py [hunk 5]
```diff
@@ -234,20 +234,17 @@ def list_project_stacks(
...
-            project_name_or_id=project_name_or_id,
-            user_name_or_id=auth_context.user.id,
-            component_id=component_id,
-            is_shared=False,
-            name=name,
-        )
-        stacks += own_stacks
+    if is_shared is None or not is_shared:
+        # only private stacks of the authenticated user can be accessed
+        if user_name_or_id and user_name_or_id == auth_context.user.id:
```
Comment (fa9r) [inline src/zenml/zen_server/routers/projects_endpoints.py:239]: Still won't work (I think) 😅 

```suggestion
        if not user_name_or_id or user_name_or_id == auth_context.user.id:
```

Comment (fa9r) [reply src/zenml/zen_server/routers/projects_endpoints.py:239]: same below ^^

**Review by fa9r — CHANGES_REQUESTED** [review]

**Review by fa9r — APPROVED** [review]
MSSVA!

**Review by bcdurak — APPROVED** [review]
Good catch and fix! Everything looks good.

**Pull request merged**